lib/Backend/NativeEntryPointData.cpp

```diff
@@ -0,0 +1,568 @@
+}
+
+void
+NativeEntryPointData::PinTypeRefs(Recycler * recycler, size_t jitPinnedTypeRefCount, void ** jitPinnedTypeRefs)
+{
+    this->runtimeTypeRefs = RecyclerNewArray(recycler, Field(void*), jitPinnedTypeRefCount + 1);
+    //js_memcpy_s(this->runtimeTypeRefs, jitPinnedTypeRefCount * sizeof(void*), jitPinnedTypeRefs, jitPinnedTypeRefCount * sizeof(void*));
+    for (size_t i = 0; i < jitPinnedTypeRefCount; i++)
+    {
+        this->runtimeTypeRefs[i] = jitPinnedTypeRefs[i];
+    }
+    this->runtimeTypeRefs[jitPinnedTypeRefCount] = nullptr;
+}
+
+PropertyGuard* NativeEntryPointData::RegisterSharedPropertyGuard(Js::PropertyId propertyId, ScriptContext* scriptContext)
+{
+    if (this->sharedPropertyGuards == nullptr)
+    {
+        Recycler* recycler = scriptContext->GetRecycler();
+        this->sharedPropertyGuards = RecyclerNew(recycler, SharedPropertyGuardDictionary, recycler);
+    }
+
+    PropertyGuard* guard = nullptr;
+    if (!this->sharedPropertyGuards->TryGetValue(propertyId, &guard))
+    {
+        ThreadContext* threadContext = scriptContext->GetThreadContext();
+        guard = threadContext->RegisterSharedPropertyGuard(propertyId);
+        this->sharedPropertyGuards->Add(propertyId, guard);
+    }
+    return guard;
+}
+
+Js::PropertyId* NativeEntryPointData::GetSharedPropertyGuards(Recycler * recycler, _Out_ unsigned int& count)
+{
+    Js::PropertyId* sharedPropertyGuards = nullptr;
+    unsigned int guardCount = 0;
+
+    if (this->sharedPropertyGuards != nullptr)
+    {
+        const unsigned int sharedPropertyGuardsCount = (unsigned int)this->sharedPropertyGuards->Count();
+        Js::PropertyId* guards = RecyclerNewArray(recycler, Js::PropertyId, sharedPropertyGuardsCount);
+        auto sharedGuardIter = this->sharedPropertyGuards->GetIterator();
+
+        while (sharedGuardIter.IsValid())
+        {
+            AnalysisAssert(guardCount < sharedPropertyGuardsCount);
+            guards[guardCount] = sharedGuardIter.CurrentKey();
+            sharedGuardIter.MoveNext();
+            ++guardCount;
+        }
+        AnalysisAssert(guardCount == sharedPropertyGuardsCount);
+
+        sharedPropertyGuards = guards;
+    }
+
+    count = guardCount;
+    return sharedPropertyGuards;
+}
+
+bool NativeEntryPointData::TryGetSharedPropertyGuard(Js::PropertyId propertyId, Js::PropertyGuard*& guard)
+{
+    return this->sharedPropertyGuards != nullptr ? this->sharedPropertyGuards->TryGetValue(propertyId, &guard) : false;
+}
+
+EquivalentTypeCache *
+NativeEntryPointData::EnsureEquivalentTypeCache(int guardCount, ScriptContext * scriptContext, EntryPointInfo * entryPointInfo)
+{
+    Assert(this->equivalentTypeCacheCount == 0);
+    Assert(this->equivalentTypeCaches == nullptr);
+
+    // Create an array of equivalent type caches on the entry point info to ensure they are kept
+    // alive for the lifetime of the entry point.
+
+    // No need to zero-initialize, since we will populate all data slots.
```

Comment:
This comment says we dont zero initialize, but the code uses a zeroing allocator- reconcile?